node/src/Pos/Delegation/Logic/Mempool.hs

```diff
@@ -189,64 +188,48 @@ processProxySKHeavyInternal psk = do
     let msg = Right psk
         consistent = verifyPsk psk
         iPk = pskIssuerPk psk
-        dPk = pskDelegatePk psk
-        -- We don't check stake for revoking certs. You can revoke
-        -- even if you don't have money anymore.
-        enoughStake = isRevokePsk psk || addressHash iPk `HS.member` richmen
-        omegaCorrect = headEpoch == pskOmega psk
-
-    -- DB related checks
-    alreadyPosted <- GS.isIssuerPostedThisEpoch $ addressHash iPk
-    pskFromDB <- GS.getPskByIssuer (Left $ pskIssuerPk psk)
-    let hasPskInDB = isJust pskFromDB
 
     -- Retrieve psk pool and perform another db check. It's
     -- guaranteed that pool is not changed when we're under
     -- 'withStateLock' lock.
     cyclePool <- runDelegationStateAction $ use dwProxySKPool
-    producesCycle <-
-        -- This is inefficient. Consider supporting this map
-        -- in-memory or changing mempool key to stakeholderId.
-        let cedeModifier =
-                HM.fromList $
-                map (bimap addressHash pskToDlgEdgeAction) $
-                HM.toList cyclePool
-        in evalMapCede cedeModifier $ detectCycleOnAddition psk
+
+    -- This is inefficient. Consider supporting this map
+    -- in-memory or changing mempool key to stakeholderId.
+    let _cmPskMods = HM.fromList $
+            map (bimap addressHash pskToDlgEdgeAction) $
+            HM.toList cyclePool
+        -- Not used since we can't have more than one psk per issuer
+        -- in mempool and "has posted this epoch" is fully backed up
+        -- by the database.
```

Comment:
I am not sure it's ok. What we if have a PSK from `x` and we are processing another PSK from `x`? I might be missing something, but it seems that verification will be considered successful, but PSK won't be inserted. Which is not too bad, but it will, for example, make this PSK relayed to others while it probably shouldn't. Am I right?

Reply:
No-no, you're missing the point. This comment says that having an extra layer of modifications over database is redudantant. 

Say that `x` has posted in this epoch before current slot. Now `x` tries to add one more cert. Checking function will understand that he did post cert using DB only. Same applies if `x` didn't post the cert. Mempool corresponds to the single block modifications and single block can't have duplicates.

Reply:
Can you clarify this check:
>         existsSameMempool <- uses dwProxySKPool $ \m -> HM.lookup iPk m == Just psk

Why do we check equality here? What if someone whose PSK is in mempool sends another PSK? Will we accept it? Doesn't it make possible for someone to spam with heavyweight PSKs which will be relayed again and again?

Reply:
@volhovM ↑

Reply:
>What if someone whose PSK is in mempool sends another PSK? Will we accept it?

We will and it will overwrite previous one with the same issuerPk.

>Doesn't it make possible for someone to spam with heavyweight PSKs which will be relayed again and again?

We have caching for it. Also "someone" should have enough stake.

Reply:
Well, AFAIU, cache doesn't matter, because it will lead to big number of messages anyway.

Regarding «have enough stake»: well, an adversary with 5% of stake can already spam the network by sending PSKs again and again from multiple machines. I suppose it's possible to make so big traffic of PSKs that sending txs will become barely possible. I don't think it's good.